PrimeNumberTheoremAnd/ResidueCalcOnRectangles.lean

```diff
@@ -228,6 +229,269 @@ that the inner square is strictly contained in the big rectangle.)
+  IntervalIntegrable f volume a b := by
+  let g : ℝ → A := fun _ ↦ 1
+  have : IntervalIntegrable g volume a b := intervalIntegrable_const
+  have := IntervalIntegrable.mul_continuousOn this hf
+  simp at this
+  exact this
+
+theorem continuous_of_inverse_continuous_nonzero (a b : ℝ) (f : ℝ → ℂ) (hf : ContinuousOn f [[a,b]]) (h : ∀ y ∈ [[a,b]], f y ≠ 0) :
+  ContinuousOn (fun x ↦ (f x)⁻¹) [[a, b]] := by
+  apply ContinuousOn.inv₀ hf h
+
+theorem ResidueTheoremAtOrigin_aux1c (a b : ℝ) :
+    let f : ℝ → ℂ := fun y => (y + I)⁻¹
+    IntervalIntegrable f volume a b := by
+  intro f
+  have : ContinuousOn f [[a, b]] := by
+    simp
+    apply continuous_of_inverse_continuous_nonzero a b (fun y ↦ (y + I)) _
+    · simp only [ne_eq, inv_eq_zero, Subtype.forall]
+      intro x _
+      by_contra h
+      have : (x + I).im = 1 := by
+        simp only [add_im, ofReal_im, I_im, zero_add]
+      rw [h] at this
+      absurd this
+      norm_num
+    apply ContinuousOn.add _ _
+    · exact Continuous.continuousOn (IsROrC.continuous_ofReal (K := ℂ))
+    exact continuousOn_const
+  exact integrable_of_continuous a b ℂ f this
+
+theorem ResidueTheoremAtOrigin_aux1c' (a b : ℝ) :
+    let f : ℝ → ℂ := fun y => (↑y + -I)⁻¹
+    IntervalIntegrable f volume a b := by
+  intro f
+  have : ContinuousOn f [[a, b]] := by
+    simp
+    apply continuous_of_inverse_continuous_nonzero a b (fun y ↦ (y + -I)) _
+    · simp only [ne_eq, inv_eq_zero, Subtype.forall]
+      intro x _
+      by_contra h
+      have : (x + -I).im = -1 := by
+        simp only [add_im, ofReal_im, neg_im, I_im, zero_add]
+      rw [h] at this
+      absurd this
+      norm_num
+    apply ContinuousOn.add _ _
+    · exact Continuous.continuousOn (IsROrC.continuous_ofReal (K := ℂ))
+    exact continuousOn_const
+  exact integrable_of_continuous a b ℂ f this
+
+theorem ResidueTheoremAtOrigin_aux1 :
+  (∫ (x : ℝ) in (-1 - 0)..(1 + 0), 1 / (x + (-0 - 1 : ℝ) * I)) -
+    ∫ (x : ℝ) in (-1 - 0)..(1 + 0), 1 / (x + (0 + 1 : ℝ) * I) = π * I
+  := by
+  simp only [neg_zero, zero_sub, ofReal_neg, ofReal_one, neg_mul, one_mul, one_div, sub_zero,
+    add_zero, zero_add]
+  rw [← intervalIntegral.integral_sub]
+  · have : ∀ x : ℝ, (x + -I)⁻¹ - (x + I)⁻¹ = (2 * I) * (1 / (1 + (x : ℝ)^2)) := by
+      intro x
+      exact ResidueTheoremAtOrigin_aux1b x
+    simp_rw [this]
+    rw [intervalIntegral.integral_const_mul (2 * I)]
+    rw [ResidueTheoremAtOrigin_aux1a]
+    simp
+    ring
+  exact ResidueTheoremAtOrigin_aux1c' (-1) 1
+  exact ResidueTheoremAtOrigin_aux1c (-1) 1
+
+theorem ResidueTheoremAtOrigin_aux2b (y : ℝ) : (1 + y * I)⁻¹ - (-1 + y * I)⁻¹ = 2 * (1 / (1 + y ^ 2)) := by
+  have hu₁ : IsUnit (1 + y * I) := by
+    apply Ne.isUnit
+    by_contra h
+    have h₁ : (1 + y * I).re = 1 := by
+      simp only [add_re, one_re, mul_re, ofReal_re, I_re, mul_zero, ofReal_im, I_im, mul_one,
+        sub_self, add_zero]
+    have h₂ : (1 + y * I).re = 0 := by
+      rw [h]
+      exact rfl
```

Comment:
maybe `rwa [h]` will close it? (tries to rewrite and close using `a`ssumptions)

Reply:
nope